### PR TITLE
Add support for `GroupBy/boundaries` queries to the ohsome API client

### LINKED ISSUE #151: Utilize ohsome API `group_by` function to make one request for multiple features
If input to OQT is a FeatureCollection or all features of a specified dataset utilize ohsome API `group by` function to only make one request for all features.

Steps for implementing.

1. `oqt.py`: Decide wether to make a batch request to ohsome API and cache resonse.
2. `ohsome.py`: Check wether a batch request have been made and data is already in cache or a request has to be made.

### PR DESCRIPTION
**Pull request by matthiasschaub**
### Description

- Support `GroupBy/boundaries` queries when input feature (boundaries) to the ohsome client is a FeatureCollection
- Add more tests for the ohsome client

### Corresponding issue
Closes #151 

### Checklist
- [x] I have updated my branch to `main` (e.g. through `git rebase main`)
- [x] My code follows the [style guide](https://github.com/GIScience/ohsome-quality-analyst/blob/main/CONTRIBUTING.md#style-guide) and was checked with [pre-commit](https://github.com/GIScience/ohsome-quality-analyst/blob/main/CONTRIBUTING.md#pre-commit) before committing
- [x] I have commented my code
- [x] I have added sufficient unit and integration [tests](https://github.com/GIScience/ohsome-quality-analyst/blob/main/docs/development_setup.md#tests)
- [x] I have updated the [CHANGELOG.md](https://github.com/GIScience/ohsome-quality-analyst/blob/main/CHANGELOG.md)
- [x] Waiting #245  
- [x] Add tests for `groupBy` queries
- [x] Add ohsome API response validation for `groupBy` and `contributions` queries
- [x] Add docstrings

### CODE REVIEW

File: workers/ohsome_quality_analyst/ohsome/client.py [hunk 1]
```diff
@@ -108,50 +111,75 @@ async def get_latest_ohsome_timestamp() -> datetime.datetime:
...
-    """Build endpoint URL of ohsome API."""
-    ohsome_api = OHSOME_API.rstrip("/")
-    if endpoint is None:
-        endpoint = layer.endpoint
-    url = ohsome_api + "/" + endpoint.rstrip("/")
+    group_by: bool = False,
+    contributions: bool = False,
+):
+    if contributions:
+        return OHSOME_API.rstrip("/") + "/contributions/latest/count"
```
Comment (matthiasschaub) [inline workers/ohsome_quality_analyst/ohsome/client.py:118]: Currently only "contributions/latest/count" is in use. Should this function be more generell or is this okay. For the latter one there needs to be more docs or a comment.

**Comment by matthiasschaub** [comment]
@joker234 I would be happy to get an review for this PR from you. The PR is no quite finished as you can detect from the open task. But those refer only to the documentation not to the code implementation of the feature. That one is finished.

If you prefer you can of course wait until I wrote down documentation (in form of docstrings and comments).

Thanks as always!

File: workers/ohsome_quality_analyst/ohsome/client.py [hunk 2]
```diff
@@ -97,61 +122,89 @@ async def query_ohsome_api(url: str, data: dict, headers: dict = {}) -> dict:
...
-        resp = await client.get(url, headers=headers)
-    timestamp_str = str(resp.json()["extractRegion"]["temporalExtent"]["toTimestamp"])
-    timestamp = datetime.datetime.strptime(timestamp_str, "%Y-%m-%dT%H:%MZ")
-    return timestamp
+        resp = await client.get(
+            url=OHSOME_API.rstrip("/") + "/metadata",
+            headers={"user-agent": USER_AGENT},
+        )
+    strtime = resp.json()["extractRegion"]["temporalExtent"]["toTimestamp"]
+    return datetime.datetime.strptime(strtime, "%Y-%m-%dT%H:%MZ")
```
Comment (joker234) [inline workers/ohsome_quality_analyst/ohsome/client.py:132]: Why did you remove the string conversion? `strptime` only takes strings. Do you make sure anywhere else, that the response only contains strings?

File: workers/ohsome_quality_analyst/ohsome/client.py [hunk 3]
```diff
@@ -97,61 +122,89 @@ async def query_ohsome_api(url: str, data: dict, headers: dict = {}) -> dict:
...
-    ohsome_api = OHSOME_API.rstrip("/")
-    if endpoint is None:
-        endpoint = layer.endpoint
-    url = ohsome_api + "/" + endpoint.rstrip("/")
+    group_by: bool = False,
+    contributions: bool = False,
+):
+    if contributions:
+        return OHSOME_API.rstrip("/") + "/contributions/latest/count"
+    url = OHSOME_API.rstrip("/") + "/" + layer.endpoint.rstrip("/")
```
Comment (joker234) [inline workers/ohsome_quality_analyst/ohsome/client.py:143]: see above (`.rstrip`)

File: workers/ohsome_quality_analyst/ohsome/client.py (hunk 1 above)
Comment (joker234) [reply workers/ohsome_quality_analyst/ohsome/client.py:142]: see comment about the variable names

File: workers/ohsome_quality_analyst/ohsome/client.py [hunk 4]
```diff
@@ -97,61 +122,89 @@ async def query_ohsome_api(url: str, data: dict, headers: dict = {}) -> dict:
...
     if ratio:
-        return url + "/" + "ratio"
+        url += "/ratio"
+    if group_by:
+        url += "/groupBy/boundary"
```
Comment (joker234) [inline workers/ohsome_quality_analyst/ohsome/client.py:147]: Not all combinations are valid. Where do you validate invalid requests?

File: workers/ohsome_quality_analyst/ohsome/client.py [hunk 5]
```diff
@@ -97,61 +122,89 @@ async def query_ohsome_api(url: str, data: dict, headers: dict = {}) -> dict:
...
     if time is not None:
         data["time"] = time
+    if ratio:
+        data["filter2"] = layer.ratio_filter
     return data
 
 
-def validate_query_results(response: dict, ratio: bool = False) -> dict:
+def validate_query_results(
+    response: dict,
```
Comment (joker234) [inline workers/ohsome_quality_analyst/ohsome/client.py:177]: The changes in this method could be more general. Too much copied code that has to be changed if anything changes in the ohsome API. Better would be a validation per parameter.

File: workers/tests/unittests/test_ohsome_client.py [hunk 6]
```diff
@@ -213,96 +151,323 @@ def test_query_layer_data_valid_2(self):
...
+                    "value": 1.0,
+                    "fromTimestamp": "2020-03-20T01:30:08.180856",
+                    "toTimestamp": "2020-04-20T01:30:08.180856",
+                }
+            ]
+        }
+        self.assertDictEqual(data, ohsome_client.validate_query_results(data))
+
+    def test_invalid_empyt(self):
         with self.assertRaises(SchemaError):
```
Comment (joker234) [inline workers/tests/unittests/test_ohsome_client.py:312]: typo

File: workers/tests/unittests/test_ohsome_client.py [hunk 7]
```diff
@@ -213,96 +151,323 @@ def test_query_layer_data_valid_2(self):
...
             ohsome_client.validate_query_results({})
 
-    def test_validate_query_results_invalid_empyt_list(self):
+    def test_invalid_empyt_list(self):
         with self.assertRaises(SchemaError):
```
Comment (joker234) [inline workers/tests/unittests/test_ohsome_client.py:316]: typo

File: workers/tests/unittests/test_ohsome_client.py [hunk 8]
```diff
@@ -213,96 +151,323 @@ def test_query_layer_data_valid_2(self):
...
-                        "ratio": "NaN",
-                        "value": 1.0,
-                        "value2": 1.0,
-                        "timestamp": "2020-03-20T01:30:08.180856",
-                    }
-                ]
-            },
-            ratio=True,
+    def test_invalid_empyt_group_by(self):
+        with self.assertRaises(SchemaError):
```
Comment (joker234) [inline workers/tests/unittests/test_ohsome_client.py:351]: typo

File: workers/tests/unittests/test_ohsome_client.py [hunk 9]
```diff
@@ -213,96 +151,323 @@ def test_query_layer_data_valid_2(self):
...
+            ohsome_client.validate_query_results({}, group_by=True)
+
+    def test_invalid_empyt_list_group_by(self):
+        with self.assertRaises(SchemaError):
```
Comment (joker234) [inline workers/tests/unittests/test_ohsome_client.py:355]: typo

File: workers/tests/unittests/test_ohsome_client.py [hunk 10]
```diff
@@ -213,96 +151,323 @@ def test_query_layer_data_valid_2(self):
...
+                    "value2": 1.0,
+                    "timestamp": "2020-03-20T01:30:08.180856",
+                }
+            ]
+        }
+        self.assertDictEqual(
+            data, ohsome_client.validate_query_results(data, ratio=True)
+        )
+
+    def test_invalid_empyt_ratio(self):
```
Comment (joker234) [inline workers/tests/unittests/test_ohsome_client.py:406]: typo

File: workers/tests/unittests/test_ohsome_client.py [hunk 11]
```diff
@@ -213,96 +151,323 @@ def test_query_layer_data_valid_2(self):
...
         with self.assertRaises(SchemaError):
             ohsome_client.validate_query_results({}, ratio=True)
 
-    def test_validate_query_results_invalid_empyt_list_ratio(self):
+    def test_invalid_empyt_list_ratio(self):
```
Comment (joker234) [inline workers/tests/unittests/test_ohsome_client.py:410]: typo

File: workers/tests/unittests/test_ohsome_client.py [hunk 12]
```diff
@@ -213,96 +151,323 @@ def test_query_layer_data_valid_2(self):
...
+                    ],
+                    "groupByObject": "foo",
+                }
+            ]
+        }
+        self.assertDictEqual(
+            data, ohsome_client.validate_query_results(data, ratio=True, group_by=True)
+        )
+
+    def test_invalid_empyt_ratio_group_by(self):
```
Comment (joker234) [inline workers/tests/unittests/test_ohsome_client.py:446]: typo

File: workers/tests/unittests/test_ohsome_client.py [hunk 13]
```diff
@@ -213,96 +151,323 @@ def test_query_layer_data_valid_2(self):
...
+        with self.assertRaises(SchemaError):
+            ohsome_client.validate_query_results({}, ratio=True, group_by=True)
+
+    def test_invalid_empyt_list_ratio_group_by(self):
```
Comment (joker234) [inline workers/tests/unittests/test_ohsome_client.py:450]: typo

**Review by joker234 — COMMENTED** [review]
Some small remarks.

File: workers/ohsome_quality_analyst/ohsome/client.py [hunk 14]
```diff
@@ -35,58 +32,86 @@ async def query(layer) -> dict:
...
+    """Query ohsome API with given Layer definition and arguments.
 
-    Time is one or more ISO-8601 conform timestring(s).
-    https://docs.ohsome.org/ohsome-api/v1/time.html
+    Args:
+        layer: Layer definition with ohsome API endpoint and parameters.
+        bpolys: Feature for a single bounding (multi)polygon.
+            FeatureCollection for "group by boundaries" queries. In this case the
+            argument 'group_by' needs to be set to 'True'.
+        time: One or more ISO-8601 conform timestring(s) as accepted by the ohsome API.
```
[the remark this answers is not in the document]
Comment (matthiasschaub) [reply workers/ohsome_quality_analyst/ohsome/client.py:48]: Yes. Because I felt like having to link to the ratio, contributions and group by ohsome API docs as well.

[the remark this answers is not in the document]
Comment (matthiasschaub) [reply workers/ohsome_quality_analyst/ohsome/client.py:48]: Happy to revert the change though.

File: workers/ohsome_quality_analyst/ohsome/client.py [hunk 15]
```diff
@@ -35,58 +32,86 @@ async def query(layer) -> dict:
...
+        ratio: Ratio of OSM elements. The Layer definition needs to have a second
+            filter defined.
+        group_by: Group by boundary.
+        contributions:  Count of the latest contributions provided to the OSM data.
```
[the remark this answers is not in the document]
Comment (matthiasschaub) [reply workers/ohsome_quality_analyst/ohsome/client.py:52]: I renamed the variable names to be very explicit.

I added a note that those parameter are a boolean flag. Initialy I hoped that would be clear by looking at the type hints.

File: workers/ohsome_quality_analyst/ohsome/client.py [hunk 16]
```diff
@@ -35,58 +32,86 @@ async def query(layer) -> dict:
...
+    """Validate data attached to the Layer object and return data.
+
+    Data will only be validated and returned immediately.
+    The ohsome API will not be queried.
+
+    Args:
+        layer: Layer with name, description and data attached to it.
+        bpolys: Feature for a single bounding (multi)polygon.
+            FeatureCollection for "group by boundaries" queries. In this case the
+            argument 'group_by' needs to be set to 'True'.
```
[the remark this answers is not in the document]
Comment (matthiasschaub) [reply workers/ohsome_quality_analyst/ohsome/client.py:78]: I did no know about that guideline. Thanks for pointing it out.
I use ` at the appropriate places.

File: workers/ohsome_quality_analyst/ohsome/client.py [hunk 17]
```diff
@@ -35,58 +32,86 @@ async def query(layer) -> dict:
...
+    A custom connection timeout is set since the ohsome API can take a long time to
+    send an answer (< 10 minutes).
 
-    # ohsome API response status codes are either 4xx and 5xx or 200
+    Raises:
+        OhsomeApiError: In case of 4xx and 5xx response status codes or invalid
+            response due to timeout during streaming.
+
+    """
+    async with httpx.AsyncClient(timeout=httpx.Timeout(300, read=660)) as client:
```
[the remark this answers is not in the document]
Comment (matthiasschaub) [reply workers/ohsome_quality_analyst/ohsome/client.py:103]: Added previous comment with explanation

File: workers/ohsome_quality_analyst/ohsome/client.py [hunk 18]
```diff
@@ -35,58 +32,86 @@ async def query(layer) -> dict:
...
+        resp = await client.post(
+            url,
+            data=data,
+            headers={"user-agent": USER_AGENT},
+        )
```
[the remark this answers is not in the document]
Comment (matthiasschaub) [reply workers/ohsome_quality_analyst/ohsome/client.py:108]: I moved the variable initialization of headers to the top.

File: workers/ohsome_quality_analyst/ohsome/client.py [hunk 19]
```diff
@@ -97,61 +122,89 @@ async def query_ohsome_api(url: str, data: dict, headers: dict = {}) -> dict:
...
-    headers = {"user-agent": USER_AGENT}
+    """Get latest unix timestamp from the ohsome API."""
     async with httpx.AsyncClient() as client:
-        resp = await client.get(url, headers=headers)
-    timestamp_str = str(resp.json()["extractRegion"]["temporalExtent"]["toTimestamp"])
-    timestamp = datetime.datetime.strptime(timestamp_str, "%Y-%m-%dT%H:%MZ")
-    return timestamp
+        resp = await client.get(
+            url=OHSOME_API.rstrip("/") + "/metadata",
+            headers={"user-agent": USER_AGENT},
```
[the remark this answers is not in the document]
Comment (matthiasschaub) [reply workers/ohsome_quality_analyst/ohsome/client.py:129]: `rstrip` does remove  trailing characters. Do you mean removing the `/` where the `OHSOME_API` variable is instantiated?

File: workers/ohsome_quality_analyst/ohsome/client.py (hunk 2 above)
Comment (matthiasschaub) [reply workers/ohsome_quality_analyst/ohsome/client.py:132]: No, I did not. I just thought that `toTimestmap` is always as string.

File: workers/ohsome_quality_analyst/ohsome/client.py (hunk 4 above)
Comment (matthiasschaub) [reply workers/ohsome_quality_analyst/ohsome/client.py:147]: There is no validation in code. The only validation that takes place is to make sure the manual defined layers are correct. 

The ratio endpoint is only used by the TagRatio Indicator. The TagRatio Indicator does only work for certain layer manually defined (see `definitions.py` and `INDICATOR_LAYER` variable). Those layer have to have a second filter element.

File: workers/tests/unittests/test_ohsome_client.py (hunk 6 above)
Comment (matthiasschaub) [reply workers/tests/unittests/test_ohsome_client.py:312]: Fixed

File: workers/tests/unittests/test_ohsome_client.py (hunk 7 above)
Comment (matthiasschaub) [reply workers/tests/unittests/test_ohsome_client.py:316]: Fixed

File: workers/tests/unittests/test_ohsome_client.py (hunk 8 above)
Comment (matthiasschaub) [reply workers/tests/unittests/test_ohsome_client.py:351]: Fixed

File: workers/tests/unittests/test_ohsome_client.py (hunk 9 above)
Comment (matthiasschaub) [reply workers/tests/unittests/test_ohsome_client.py:355]: Fixed

File: workers/tests/unittests/test_ohsome_client.py (hunk 10 above)
Comment (matthiasschaub) [reply workers/tests/unittests/test_ohsome_client.py:406]: Fixed

File: workers/tests/unittests/test_ohsome_client.py (hunk 11 above)
Comment (matthiasschaub) [reply workers/tests/unittests/test_ohsome_client.py:410]: Fixed

File: workers/tests/unittests/test_ohsome_client.py (hunk 12 above)
Comment (matthiasschaub) [reply workers/tests/unittests/test_ohsome_client.py:446]: Fixed

File: workers/tests/unittests/test_ohsome_client.py (hunk 13 above)
Comment (matthiasschaub) [reply workers/tests/unittests/test_ohsome_client.py:450]: Fixed

File: workers/ohsome_quality_analyst/ohsome/client.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (joker234) [reply workers/ohsome_quality_analyst/ohsome/client.py]: Thanks, I think the variable names are much better now. I'm not sure about the `(boolean flag)` note. As you mentioned this is already in the type hints, so it is probably not necessary. I think you could remove the notes now (with the more comprehensible variable names).

File: workers/ohsome_quality_analyst/ohsome/client.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (joker234) [reply workers/ohsome_quality_analyst/ohsome/client.py:48]: No, that's fine.

File: workers/ohsome_quality_analyst/ohsome/client.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (joker234) [reply workers/ohsome_quality_analyst/ohsome/client.py]: It's not about `rstrip` itself, but why you would use that at *every* usage of `OHSOME_API`. Wouldn't it make more sense to do the `rstrip` once in [`definitons.py`](https://github.com/GIScience/ohsome-quality-analyst/blob/208e27bc0302122348093f2a7aff23d439fa0b25/workers/ohsome_quality_analyst/utils/definitions.py#L126) or just assume that the definition is without a trailing `/`?

File: workers/ohsome_quality_analyst/ohsome/client.py (hunk 1 above)
Comment (joker234) [reply workers/ohsome_quality_analyst/ohsome/client.py:140]: I think with the changed variable names this is okay for now.

**Review by joker234 — COMMENTED** [review]
:+1: only two comments left

File: workers/ohsome_quality_analyst/ohsome/client.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (matthiasschaub) [reply workers/ohsome_quality_analyst/ohsome/client.py]: Got you. Thanks for pointing this out. I fixed it by running `rstrip` once in `definition.py`.

File: workers/ohsome_quality_analyst/ohsome/client.py (hunk 3 above)
Comment (matthiasschaub) [reply workers/ohsome_quality_analyst/ohsome/client.py:141]: fixed

**Review by joker234 — DISMISSED** [review]

File: workers/ohsome_quality_analyst/ohsome/client.py (hunk 5 above)
Comment (joker234) [reply workers/ohsome_quality_analyst/ohsome/client.py:175]: see 3b4f51c3d5d7d8e5857ccaa5cdd67328f244fb2f

File: workers/ohsome_quality_analyst/ohsome/client.py (hunk 4 above)
Comment (matthiasschaub) [reply workers/ohsome_quality_analyst/ohsome/client.py]: https://github.com/GIScience/ohsome-quality-analyst/issues/300

**Review by joker234 — APPROVED** [review]
